### PR TITLE
Subhalo mass function given parent halo

### LINKED ISSUE #148: Subhalos in parent halos
## Description

We need a function to add dark matter subhalo to a parent halo (section three of Vale and Ostriker 2005). Possibly related to #177.

## Inputs
 - Parent halo properties (mostly mass, but possibly also redshift).

## Outputs
 - List of subhalo properties.

## References
- [Vale and Ostriker, 2005, astro-ph/0511816.](https://arxiv.org/pdf/astro-ph/0511816.pdf)

### PR DESCRIPTION
**Pull request by Lucia-Fonseca**
## Description
This function implements the distribution of the subhaloes, based on the subhalo mass function (SHMF), N(m|M)dm, which gives the num-ber of subhaloes in the mass rangemtom+dmfor a parent halo of mass M. This is described by equations (3) and (4) in [1]. This pull request resolves #148.
## Checklist
- [x] Follow the [Contributor Guidelines](https://github.com/skypyproject/skypy/blob/master/CONTRIBUTING.md)
- [x] Write unit tests
- [x] Write documentation strings
- [x] Assign someone from your working team to review this pull request
- [x] Assign someone from the infrastructure team to review this pull request

### CODE REVIEW

**Comment by Lucia-Fonseca** [comment]
This is ready for review

**Comment by aamara** [comment]
This looks really good. On the tests, it looks like the only thing that fails is codeclimate, which seems to only want four inputs. I don't know how seriously we're taking this. 

The other thought I had is that at the moment the tests only assert the lengths of the arrays. Is there a way to add a couple more asserts to make sure values are consistent. This is probably harder for the sampler, but maybe can be done for the mass function code?

**Comment by Lucia-Fonseca** [comment]
Some tests need to be changed

File: skypy/halo/mass.py [hunk 1]
```diff
@@ -314,3 +320,161 @@ def _sigma_squared(M, k, Pk, growth_function, cosmology):
...
+    References
+    ----------
+    .. [1] Vale, A. and Ostriker, J.P. (2005), arXiv: astro-ph/0511816.
+    '''
+
+    if model == 'Vale':
+        subhalo_fraction0, alpha, beta, mcut = params
+        A = _subhalo_amplitude(halo_mass, params)
+    else:
+        A, alpha, beta = params
```
Comment (rrjbca) [inline skypy/halo/mass.py:450]: For now lets just implement the V&O subhalo model so make `subhalo_fraction0, alpha, beta, mcut` arguments of the function instead of using `params`

File: skypy/halo/mass.py [hunk 2]
```diff
@@ -314,3 +320,161 @@ def _sigma_squared(M, k, Pk, growth_function, cosmology):
...
+def _subhalo_amplitude(M, params):
+    subhalo_fraction0, alpha, beta, mcut = params
+    x_cut = mcut / (beta * M)
+
+    return (subhalo_fraction0 / beta) / \
+        gamma(2.0 - alpha) * (1.0 - gammainc(2.0 - alpha, x_cut))
+
+
+def subhalo_mass_sampler(m_min, m_max, resolution,
+                         halo_mass, params, model=None):
```
Comment (rrjbca) [inline skypy/halo/mass.py:394]: Would it make sense here to use `m_max = halo_mass`? i.e. subhalos cannot be more massive than their parents

File: skypy/halo/mass.py [hunk 3]
```diff
@@ -314,3 +320,161 @@ def _sigma_squared(M, k, Pk, growth_function, cosmology):
...
+        A = _subhalo_amplitude(halo_mass, params)
+    else:
+        A, alpha, beta = params
+
+    x = M / (beta * halo_mass)
+
+    return A * np.power(x, -alpha) * np.exp(-x) / (beta * halo_mass)
+
+
+def _subhalo_amplitude(M, params):
```
Comment (rrjbca) [inline skypy/halo/mass.py:385]: Reference astro-ph/0511816 Equation 4 in a comment

File: skypy/halo/tests/test_mass.py [hunk 4]
```diff
@@ -59,7 +59,7 @@ def test_halo_mass_function():
     assert allclose(array_output_PS, PS_massf)
 
 
-def halo_mass_sampler():
+def test_halo_mass_sampler():
```
Comment (rrjbca) [inline skypy/halo/tests/test_mass.py:62]: Again, these fixes have nothing to do with the subhalo mass sampler so put them in a separate pull request

**Review by rrjbca — CHANGES_REQUESTED** [review]

File: skypy/halo/mass.py (hunk 2 above)
Comment (Lucia-Fonseca) [reply skypy/halo/mass.py:394]: That's a good question. Subhalos are not more massive than their parents but their maximum mass is not their parent's mass either, right? I could raise an error if `m_max > halo_mas`

File: skypy/halo/mass.py [hunk 5]
```diff
@@ -314,3 +320,161 @@ def _sigma_squared(M, k, Pk, growth_function, cosmology):
...
+    '''
+    if model == 'Vale':
+        subhalo_fraction0, alpha, beta, mcut = params
+        A = _subhalo_amplitude(halo_mass, params)
+    else:
+        A, alpha, beta = params
+
+    x = M / (beta * halo_mass)
+
+    return A * np.power(x, -alpha) * np.exp(-x) / (beta * halo_mass)
```
[the remark this answers is not in the document]
Comment (Lucia-Fonseca) [reply skypy/halo/mass.py:382]: Agreed

**Comment by Lucia-Fonseca** [comment]
I am using `skypy.utils.random.schechter`, and I get this error when I run the tests locally on my computer: `TypeError: schechter() got an unexpected keyword argument 'scale'`.  Any idea?

**Comment by rrjbca** [comment]
The PR branch is not up to date with master. Click 'Resolve conflicts', resolve the merge conflict and then pull the updated branch. Or fetch master and merge the changes locally.

File: skypy/halo/mass.py (hunk 2 above)
Comment (rrjbca) [reply skypy/halo/mass.py]: No, we will be sampling perhaps many millions of subhalos using this function we can't afford it to crash if one doesn't satisfy your inequality. I suggest you ignore my suggestion and leave it as it is for now

**Review by rrjbca — CHANGES_REQUESTED** [review]
Some code refactoring needed, see suggestions. Also a couple of questions:
- gamma_M is defined as an integral from zero to m_cut, but we are sampling from m_min to m_max(=m_cut?) do we need to redefine the limits of the integral defining gamma_M? Or does e.g. the line `n_subhalos = _subhalo_amplitude(M, params) *  gammaincc(1.0 - alpha, x_low) * gamma(1.0 - alpha)` correctly account for this?
- Reading sections 3.1 and 3.2 it appears that m_cut is not an independent variable, rather it is determined by the mass loss of the system. Can we understand and implement this?

**Comment by ntessore** [comment]
Just as a heads up, it is almost certainly better to replace the calls to `gamma` and `gammaincc` by a small numerical integration, as e.g. done in #244. As it stands you need to work around integer values of `alpha`.

**Comment by Lucia-Fonseca** [comment]
@rrjbca Thank you for your comments.
> * gamma_M is defined as an integral from zero to m_cut, but we are sampling from m_min to m_max(=m_cut?) do we need to redefine the limits of the integral defining gamma_M? Or does e.g. the line `n_subhalos = _subhalo_amplitude(M, params) *  gammaincc(1.0 - alpha, x_low) * gamma(1.0 - alpha)` correctly account for this?

- You are right with respect to `m_cut`, that would correspond to the most massive object that you would call "subhalo". That is, the user would be sampling from m_min to m_max(=m_cut). 
- n_subhalos is the integral of equation 3 from m_min to infinity. Perhaps you are correct and the upper limit should be m_cut.

> * Reading sections 3.1 and 3.2 it appears that m_cut is not an independent variable, rather it is determined by the mass loss of the system. Can we understand and implement this?
Indeed you are correct here too. I quote the paper "[m_cut(M)] This value will be set by assuming that the maximum original mass is half the total mass of the parent, M/2".
If I understand this correctly, `m_max` or `m_cut` should not be given by the user but fixed to be `halo_mass / 2`.

**Comment by Lucia-Fonseca** [comment]
> Just as a heads up, it is almost certainly better to replace the calls to `gamma` and `gammaincc` by a small numerical integration, as e.g. done in #244. As it stands you need to work around integer values of `alpha`.

Agreed

File: skypy/halo/mass.py [hunk 6]
```diff
@@ -312,3 +318,99 @@ def _sigma_squared(M, k, Pk, growth_function, cosmology):
...
+
+        # Sample from the Schechter function
+        x_min = np.min(m) / m_star
+        x_max = np.max(m) / m_star
+
+        schechter_sampling = A * schechter(alpha, x_min, x_max, resolution,
+                                           size=int(nsh[i]), scale=m_star)
+
+        subhalo_list.append(schechter_sampling)
+        i += 1
```
[the remark this answers is not in the document]
Comment (Lucia-Fonseca) [reply skypy/halo/mass.py]: Great, thanks @rrjbca . 
Q: I do not see why I am not supposed to multiply by `A`, the `subhalo_amplitude`. `schechter` is the normalised pdf, and in equation 3 `A` is the amplitude of the Schechter function. Is there anything I am missing here?

**Comment by rrjbca** [comment]
>   Indeed you are correct here too. I quote the paper "[m_cut(M)] This value will be set by assuming that the maximum original mass is half the total mass of the parent, M/2".
>   If I understand this correctly, `m_max` or `m_cut` should not be given by the user but fixed to be `halo_mass / 2`.

Not quite, it says the maximum "original" mass of the subhalo is M/2, which I think is different from the "present day" or "evolved" mass. The original mass and present day mass are related by the "mass loss factor". I need to read section 3.1 more carefully to understand how we incorporate mass loss.

**Comment by Lucia-Fonseca** [comment]
Tests still fail due to `test_quenching.py` failing.

**Comment by Lucia-Fonseca** [comment]
All checks passed!

File: skypy/halo/mass.py [hunk 7]
```diff
@@ -312,3 +322,133 @@ def _sigma_squared(M, k, Pk, growth_function, cosmology):
...
+    return gamma_M / (beta * gamma(2.0 - alpha))
+
+
+def number_subhalos(halo_mass, alpha, beta, gamma_M, x, m_min):
+    r'''Number of subhalos.
+    This function calculates the number of subhalos for a parent halo of given mass
+    according to the model of Vale & Ostriker  2004 [1]_. The mean number of subhalos
+    above a mass threshold can be obtained by integrating equation (1) in [1]. The
+    number of subhalos returned is randomly drawn from a Poisson distribution with
+    that mean.
```
Comment (rrjbca) [inline skypy/halo/mass.py:338]:
```suggestion
    This function calculates the number of subhalos above a given initial mass
    for a parent halo of given mass according to the model of Vale & Ostriker
    2004 [1]_ equation (7). The number of subhalos returned can optionally be
    sampled from a Poisson distribution with that mean.
```

File: skypy/halo/mass.py [hunk 8]
```diff
@@ -312,3 +322,133 @@ def _sigma_squared(M, k, Pk, growth_function, cosmology):
...
+        Current stripped mass is given by :math:`x m_{min}`.
+    resolution: int
+        Resolution of the inverse transform sampling spline.
+
+    Returns
+    --------
+    sample: (nh, ) array_like
+        List of samples drawn from the mass function, in units of solar mass.
+        The length corresponds to the addition of all subhalos of all parent halos,
+        i.e. `np.sum(nsubhalos)`
```
Comment (rrjbca) [inline skypy/halo/mass.py:422]:
```suggestion
        List of original masses drawn from the subhalo mass function for each
        parent halo, in units of solar mass. The length corresponds to the total
        number of subhalos for all parent halos, i.e. `np.sum(nsubhalos)`
```

File: skypy/halo/mass.py [hunk 9]
```diff
@@ -312,3 +322,133 @@ def _sigma_squared(M, k, Pk, growth_function, cosmology):
...
+    # with mean n_subhalos
+    return np.random.poisson(n_subhalos)
+
+
+def subhalo_mass_sampler(halo_mass, nsubhalos, alpha, beta, gamma_M,
+                         x, m_min, resolution):
+    r'''Subhalo mass sampler.
+    This function samples subhaloes from their mass function, given the mass
+    of the parent halo.
+    This function uses equation 1 in [1]_ and the upper subhalo mass limit from [2]_.
```
Comment (rrjbca) [inline skypy/halo/mass.py:395]:
```suggestion
    r'''Subhalo mass sampler.
    This function samples the original, unstriped masses of subhaloes from the
    subhalo mass function of their parent halo with a constant mass stripping
    factor given by equation 1 in [1]_ and the upper subhalo mass limit from
    [2]_.
```

**Review by rrjbca — CHANGES_REQUESTED** [review]
Need tests for both number_subhalos and subhalo_mass_sampler. Suggest:
- Choose a special case for the parameters such that the gamma functions and hence the number of subhalos is analytic. Check the analytic solution for noise=False
- For each halo check that each of it's subhalos in array_output satisfies m_min < m < m_max

File: skypy/halo/mass.py [hunk 10]
```diff
@@ -312,3 +322,133 @@ def _sigma_squared(M, k, Pk, growth_function, cosmology):
...
+    ----------
+    .. [1] Vale, A. and Ostriker, J.P. (2004), arXiv: astro-ph/0402500.
+    .. [2] Vale, A. and Ostriker, J.P. (2004), arXiv: astro-ph/0511816.
+    '''
+    halo_mass = np.atleast_1d(halo_mass)
+    nsubhalos = np.atleast_1d(nsubhalos)
+    subhalo_list = []
+    for M, n in zip(halo_mass, nsubhalos):
+        x_min = m_min / (x * beta * M)
+        x_max = 0.5 * M / (x * beta * M)
```
[the remark this answers is not in the document]
Comment (Lucia-Fonseca) [reply skypy/halo/mass.py:451]:
```suggestion
        x_max = 0.5 * halo_mass / (x * beta * M)
```

**Review by rrjbca — CHANGES_REQUESTED** [review]
In your test for subhalo_mass sampler, halo_mass and nsh are floats. It would be good to repeat the test but where they are arrays e.g:
```
halo_mass = np.array([1.0e12, 1.0e14])
nsh = np.array([10, 100])
...
...
...
assert len(array_output) == np.sum(nsh)
assert np.all(sh_mass[:10] > shm_min) and np.all(sh_mass[:10] < 0.5 * halo_mass[0])
assert np.all(sh_mass[10:] > shm_min) and np.all(sh_mass[10:] < 0.5 * halo_mass[1])

**Review by rrjbca — APPROVED** [review]
